### PR TITLE
Idiomatic helm chart for Teleport

### PR DESCRIPTION
**Pull request by mumoshu**
Hey, thanks a lot for maintaining the awesome project!

Here is an (hopefully) idiomatic helm chart for Teleport, with instructions to test it locally using minikube + ngrok.

I was just trying to run Teleport on Kubernetes for #1986 and thought a more idiomatic helm chart would be helpful as a foundation towards the upcoming 2.7.0 release. I'm sending this PR because I was unable to just wait for it :)

Please feel free to leave questions, comments, requests, etc.

### CODE REVIEW

File: examples/chart/teleport/templates/deployment.yaml
Comment (mumoshu) [inline examples/chart/teleport/templates/deployment.yaml]: TODO: Make this configurable. This can't be hard-coded. The user should decide whether to terminate TLS on Teleport Proxy or the load balancer in front of that. `--insecure-no-tls` should be optionally configured to be enabled only for the latter case.

File: examples/chart/teleport/templates/deployment.yaml
Comment (mumoshu) [inline examples/chart/teleport/templates/deployment.yaml]: These three values should be fully configurable via the `image:` settings in `values.yaml`.

File: examples/chart/teleport/templates/deployment.yaml
Comment (mumoshu) [inline examples/chart/teleport/templates/deployment.yaml]: You should never hard-code, or even set resource requests/limits. Make it completely up to the users, and use the idiomatic key `resources` in `values.yaml` for that.

**Comment by mumoshu** [comment]
Logging into the Teleport Proxy w/ OneLogin, both from `tsh` build from the master branch, and from the Proxy's web UI worked. I also verified that I'm able to SSH into the teleport pod running fro the web UI.

And now, I'm hitting this even after I've successfully `tsh --proxy myproxyhost --auth=onelogin login` as a remote user, and `tsh --proxy myproxyhost --auth=local --user=admin login` as a local user:

```
teleport-6f6c49b4b8-b75t4 teleport DEBU [PROXY:KUB] Returning existing creds for user: admin, groups: [system:masters], cluster: lens.gravitational.co. proxy/forwarder.go:530
teleport-6f6c49b4b8-b75t4 teleport INFO             Round trip: https://kubernetes.default.svc.cluster.local:443/api/v1/namespaces/default/pods/ngrok-6b68cb54f7-c5brc, code: 401, duration: 11.899519ms tls:version: 303, tls:resume:false, tls:csuite:cca8, tls:server:0.tcp.ngrok.io forward/fwd.go:176
```

Seems like the TLS config provided from the proxy to the K8S apiserver isn't authenticating me. Would this by any chance due to the limitation in minikube? https://github.com/kubernetes/minikube/issues/1647

**Comment by klizhentas** [comment]
Hey, keep up the great work, thanks for this PR!

**Comment by klizhentas** [comment]
regarding minikube problem - not sure, I haven't tried this. If it was the cert controller problem it would manifest itself in proxy being not able to get the certificate in the first place, but it looks like it just fails authentication with existing credentials. I can try to repro, do what version of k8s and minikube are you testing with?

**Comment by mumoshu** [comment]
@klizhentas Hi. It turned out that my minikube was up-to-date enough so it wasn't an issue.

Actually I could successfully obtain a cert with usage `client auth`, set it under the `users` section of `kubeconfig`, specify the `user` that uses the cert by `kubectl -v=6 --user=genuser version`. K8S on my minikube box authenticated me with the cert generated with Certificates API without any problem.

You can use the following steps for reproduction:

```
$ cat <<EOF | cfssl genkey - | cfssljson -bare genuser
{
  "hosts": [
    "https://192.168.99.100:8443"
  ],
  "CN": "genuser",
   "names": [{
      "O": "system:masters"
   }],
  "key": {
    "algo": "ecdsa",
    "size": 256
  }
}
EOF

$ cat <<EOF | kubectl create -f -
apiVersion: certificates.k8s.io/v1beta1
kind: CertificateSigningRequest
metadata:
  name: genuser
spec:
  groups:
  - system:masters
  request: $(cat genuser.csr | base64 | tr -d '\n')
  usages:
  - client auth
EOF

$ kubectl certificate approve genuser

$ kubectl get csr genuser -o jsonpath='{.status.certificate}'     | base64 -D > genuser.crt

$ emacs ~/.kube/config
```

Update the `users` section of `kubeconfig` to include the cert:

```
users:
- name: minikube
  user:
    client-certificate: /path/to/.minikube/client.crt
    client-key: /path/to/.minikube/client.key
# ADD THIS
- name: genuser
  user:
    client-certificate: /path/to/genuser.crt
    client-key: /path/to/genuser-key.pem
```

Specify kubectl to use the `genuser` and its corresponding client cert/key for authentication:

```
$ kubectl version -v=6 --user=genuser
I0619 10:04:15.767874   55610 loader.go:357] Config loaded from file /path/to/.kube/config
I0619 10:04:15.794904   55610 round_trippers.go:405] GET https://192.168.99.100:8443/version 200 OK in 9 milliseconds
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-05-12T04:12:12Z", GoVersion:"go1.9.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:44:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

$ kubectl get no -v=6 --user=genuser
I0619 10:29:33.110322   59166 loader.go:357] Config loaded from file /path/to/.kube/config
I0619 10:29:33.151680   59166 round_trippers.go:405] GET https://192.168.99.100:8443/api/v1/nodes?limit=500 200 OK in 10 milliseconds
I0619 10:29:33.166275   59166 round_trippers.go:405] GET https://192.168.99.100:8443/openapi/v2 200 OK in 13 milliseconds
NAME       STATUS    ROLES     AGE       VERSION
minikube   Ready     master    11d       v1.10.0
```

**Comment by mumoshu** [comment]
> I can try to repro, do what version of k8s and minikube are you testing with?

Thanks! Here's my setup:

minikube: `0.27.0`
kubectl: `1.10.2`
k8s: `1.10.0`

**Comment by mumoshu** [comment]
In case your minikube doesn't enable RBAC by default yet, perhaps this helps:

```
minikube start --cpus 2 --memory 4096  --extra-config=apiserver.authorization-mode=RBAC \
    && sleep 30 \
    && kubectl create clusterrolebinding add-on-cluster-admin --clusterrole=cluster-admin --serviceaccount=kube-system:default
```

**Comment by mumoshu** [comment]
@klizhentas  I have read a few places in the code-base to figure out what's going here. Perhaps the certs are missing the `ClientAuth` usage.

---

I believe this is the entrypoint - proxy should obtain a valid cert generated using K8S' Certificates API here.
https://github.com/gravitational/teleport/blob/0ad6c4859eda9af3000ac62b2bddefdd5d515498/lib/kube/proxy/forwarder.go#L638

auth `Client`: https://github.com/gravitational/teleport/blob/86fa0e7a72c9e6cc934aa59b8313767b607ab01d/lib/auth/clt.go#L241

`AuthWithRoles`: https://github.com/gravitational/teleport/blob/803b8e5bf9085f6fa386a0ed7ba1faf61938e3b7/lib/auth/auth_with_roles.go#L1205

and `AuthServer:` https://github.com/gravitational/teleport/blob/0ad6c4859eda9af3000ac62b2bddefdd5d515498/lib/auth/kube.go#L49

Now, I'm unsure why we're providing `UsageServerAuth ` in cert usages:
https://github.com/gravitational/teleport/blob/0ad6c4859eda9af3000ac62b2bddefdd5d515498/lib/kube/authority/authority.go#L43-L55

Shouldn't this be [`UsageClientAuth`](https://godoc.org/k8s.io/api/certificates/v1beta1#KeyUsage) by any chance?

`UsageClientAuth` would correspond in what I've included in the working example csr above:

```
usages:
  - client auth
```

AFAIK, K8S API doesn't authenticate you when your cert is missing the client auth usage, because what K8S API is doing when authenticating API clients is TLS Client Auth.

**Comment by mumoshu** [comment]
Submitted #2016 that should fix my issue theoretically. It is theoretical because I was unable to find a way to build my own docker image with the fix, and test it. Could you build one and cut 2.7.0-alpha.4 for testing?

In my understanding, the Dockerfile to build the image isn't included in this repo, almost certainly because this is OSS ver of the teleport-ent :)

**Comment by klizhentas** [comment]
@mumoshu you may want to retry with alpha-3 container, I've just replublished. I will try to set up auto publishing of container images this week, so you will be unblocked

**Comment by mumoshu** [comment]
@klizhentas Thanks. I just retried with the re-pulled image and it is working like a charm 🎉 

I basically did the following in order to test the feature in isolation:

```
# Expose teleport's kube_proxy at localhost:3026
$ kubectx minikube; kubectl port-forward $(kubectget po | grep teleport | grep Running | cut -d' ' -f1) 3026:3026

# Obtain the client key/cert and ca for the teleport's kube proxy, via the teleport proxy web exposed via ngrok, and point kubectl to the localhost:3026
$ kube_proxy_host=MY_NGROK_TUNNEL.ngrok.io bash -c 'tsh --proxy=$kube_proxy_host:443 --auth=onelogin login && sed -i.bak -e "s/server: https:\/\/$kube_proxy_host/server: https:\/\/localhost/" ~/.kube/config; kubectx $kube_proxy_host; kubectl --insecure-skip-tls-verify get no'
https://<your idp auth endpoint>

You are now logged in
Switched to context "MY_NGROK_TUNNEL.ngrok.io".
NAME       STATUS    ROLES     AGE       VERSION
minikube   Ready     master    14d       v1.10.0
```

**Comment by mumoshu** [comment]
Now, the server auth from kubectl to the Teleport proxy's kube proxy endpoint is failing like this:

```
Unable to connect to the server: x509: certificate is valid for teleport-6f6c49b4b8-lxvx9, https, not MY_NGROK_TUNNEL.ngrok.io
```

I was assuming the server cert to be issued for the hostname specified by `proxy_service.public_addr` in my `teleport.yaml`, which turns out not true, because `teleport-6f6c49b4b8-lxvx9` is the name of the teleport pod running on my minikube.

How can I set the hostname used for the cert?

**Comment by klizhentas** [comment]
you can set it as a `public_addr` parameter in the proxy section

**Comment by mumoshu** [comment]
The server cert returned by the Teleport proxy's kube endpoint:

```
$ openssl s_client -servername MY_NGROK_TUNNEL.ngrok.io -connect MY_NGROK_TUNNEL.ngrok.io:$TELE_KUBE_PROXY_PORT </dev/null | sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' | openssl x509 -text
```

contains:

```
Subject: O=Proxy, CN=7b573e4a-6bcf-4e99-be81-97467bf6d568.lens.gravitational.co
```

and

```
X509v3 Subject Alternative Name:
                DNS:teleport-6f6c49b4b8-lxvx9, DNS:https
```

The CN is something that is generated by teleport from `cluster_name` in my `teleport.yaml`, and the only alt name is the teleport pod name.
So, probably what I'd want is to add an externally accessible hostname of teleport proxies as alt. names of the cert.

**Comment by mumoshu** [comment]
@klizhentas Thanks! I do have something like this in my `teleport.yaml`:

```
proxy_service:
  enabled: yes
  public_addr: https://MY_NGROK_TUNNEL.ngrok.io/
  # snip
```

and my teleport pod is up-to-date with the yaml.

Am I missing something?
Or would there be any chance that Teleport's kube-proxy endpoint has something wrong with the cert?

**Comment by mumoshu** [comment]
So I had the `public_addr` in a wrong format. Something like `public_addr: https://anyhostname/` doesn't emit any errors, but instead end up with an empty port and a non-sense hostname `https` 😆 
You can see `DNS:https` in the subject alt. names in my cert above, that supports my theory.

**Comment by mumoshu** [comment]
Okay so after updating my `teleport.yaml` to this:

```
proxy_service:
  enabled: yes
  public_addr: MY_NGROK_TUNNEL.ngrok.io
  # snip
```

It's working. I can now authenticate via teleport's kube proxy endpoint exposed by ngrok 🎉
Note that as you can't expose multiple ports on the single ngrok endopint, you need to modify `/etc/hosts` temporarily to point the ngrok endpoint to another one serving teleport's kube proxy.

```bash
$ teleport_proxy_web_host=MY_NGROK_TUNNEL.ngrok.io \
teleport_proxy_kube_host=0.tcp.ngrok.io \
teleport_proxy_kube_hostport=MY_NGROK_TUNNEL.ngrok.io:MY_NGROK_TCP_PORT_FOR_TELEPORT \
  bash -c '
    tsh --proxy=$teleport_proxy_web_host:443 --auth=onelogin login \
    && sudo hostess add $teleport_proxy_web_host $(host $teleport_proxy_kube_host | cut -d" " -f 4) \
    && sed -i.bak -e "s/server: https:\/\/$teleport_proxy_web_host:3026/server: https:\/\/$teleport_proxy_kube_hostport/" ~/.kube/config; \
    kubectx $teleport_proxy_web_host \
    && kubectl get no \
    && sudo hostess del $teleport_proxy_web_host
  '

If browser window does not open automatically, open it by clicking on the link:
https://MY_IDP_AUTH_ENDPOINT

You are now logged in
Added MY_NGROK_TUNNEL.ngrok.io -> IP_ADDR_TO_NGROK_EXPOSED_TELEPORT_KUBE_PROXY_ENDPOINT (On)
Switched to context "MY_NGROK_TUNNEL.ngrok.io".
NAME       STATUS    ROLES     AGE       VERSION
minikube   Ready     master    14d       v1.10.0
Deleted MY_NGROK_TUNNEL.ngrok.io
```

**Comment by klizhentas** [comment]
hey, just noticed this, why the dependency on ngrok?

**Comment by mumoshu** [comment]
@klizhentas Good point! What would be your recommendation for an easiest way to test Teleport on K8S locally?

My original answer for that was the combination of SAML + Teleport on Minikube + OpenSSH sshd on an EC2 instance.

However, it was based on my misassumptions that:

1. the EC2 instance needs access to Teleport and
2. my IdP(OneLogin) doesn't allow SP on `localhost`

1 is obviously untrue. I'm not yet sure about 2 but if it's ok, maybe I'd better rewrite the steps to use `kubectl port-forward` instead.

**Comment by klizhentas** [comment]
localhost should be possible just for testing - you can always override the local DNS resolution otherwise

**Comment by mumoshu** [comment]
@klizhentas Thanks! Just tried it but using `localhost` seems to be not a complete way to test Teleport locally.

For example, `tsh play` forces you to use TLS against 3080.

```
$ tsh --proxy=localhost --insecure play c4e7ef24-7600-11e8-93db-0242ac110005
WARNING: You are using insecure connection to SSH proxy https://localhost:3080`
error: Get https://localhost:3080/v1/webapi/ping: http: server gave HTTP response to HTTPS client
```

Do I need to somehow terminate TLS between `tsh` and teleport proxy(web)?
Or can you tell `tsh` to not use TLS against teleport on `localhost`?

**Comment by mumoshu** [comment]
> you can always override the local DNS resolution otherwise

Thanks for the guidance.
Yes, it would work as similar as I've done it with `hostess` for the ngrok example.

**Comment by klizhentas** [comment]
you probably need to turn on tls on the proxy side

**Comment by mumoshu** [comment]
@klizhentas Thanks! Probably I'm a bit confusing. Would you mind reviewing the below test setup?

---

I'll start by issueing a server cert for the proxy.

> Alternatively I can just omit `--insecure-no-tls` from the `teleport` command hard-coded in the deplyoment template. But I'll proceed with issueing one myself anyway. `--insecure-no-tls` results in a self-signed cert, which fails TLS server auth. `tsh --insecure` would bypass the server auth.

The cert's subject alt. names should include a hostname `teleport.example.com`, which is later used to override the local DNS resolution. To include the hostname in the alt names, I would write `teleport.yaml` like:

```yaml
proxy_service:
  public_addr: teleport.example.com
```

Also, the cert should be signed by the well-known CA that is present in my machine so that I can omit the `--insecure` flag of `tsh` commands. The well-known CA can be one that is generated with a tool like openssl, as long as it is trusted as a root CA.

Then, I'll provide the server key/cert to the proxy via:

```yaml
proxy_service:
  https_key_file: /var/lib/certs/key
  https_cert_file: /var/lib/certs/crt
```

in my `teleport.yaml`.

**Comment by mumoshu** [comment]
@klizhentas I've successfully got it working 🎉See 2ab8b3a for details.

I'll remove all the revelant code/instructions relevant to the teleport+ngrok scenario in favor of the new way.

Do you have anything that you want me to address before finishing this?

File: examples/chart/teleport/README.md
Comment (klizhentas) [inline examples/chart/teleport/README.md]: Remove reference for ngrok here and below

**Comment by klizhentas** [comment]
I will do a pass over day-two to give you more structured feedback

File: examples/chart/teleport/config/teleport.yaml
Comment (klizhentas) [inline examples/chart/teleport/config/teleport.yaml]: extra comment

File: examples/chart/teleport/config/teleport.yaml
Comment (klizhentas) [inline examples/chart/teleport/config/teleport.yaml]: this should be a chart parameter I think

File: examples/chart/teleport/minikube.values.yaml
Comment (klizhentas) [inline examples/chart/teleport/minikube.values.yaml]: TLS_CERT_FILE -> SSL prefix is obsolete

File: examples/chart/teleport/scripts/cleanup-on-minikube
Comment (klizhentas) [inline examples/chart/teleport/scripts/cleanup-on-minikube]: this seems like a very custom script, let's remove it

File: examples/chart/teleport/scripts/create-internal-pki
Comment (klizhentas) [inline examples/chart/teleport/scripts/create-internal-pki]: same here

File: examples/chart/teleport/scripts/install-on-minikube
Comment (klizhentas) [inline examples/chart/teleport/scripts/install-on-minikube]: remove extra port forwards now that we know they are not necessary

File: examples/chart/teleport/templates/service.yaml
Comment (klizhentas) [inline examples/chart/teleport/templates/service.yaml]: it would be helpful to make this configurable, because I usually need load balancer when setting up on GKE, I imagine in many cases that's not necessary, same about ports?

**Review by klizhentas — CHANGES_REQUESTED** [review]
I've left a bunch of comments mostly regarding extra files and adding a couple of more configuration options.

File: examples/chart/teleport/scripts/install-on-minikube
Comment (mumoshu) [reply examples/chart/teleport/scripts/install-on-minikube]: At least 3080 and 3026 are required in order to test it locally via `teleport.example.com` resolved via the local resolver config. 

> remove extra port forwards

So, maybe you are talking about removing 3022, 3023, 3024 and 3025 specifically, right?

File: examples/chart/teleport/templates/service.yaml
Comment (mumoshu) [reply examples/chart/teleport/templates/service.yaml]: Makes sense! Actually I need to run Teleport on K8S on AWS so I'll make necessary changes for that, which should also address your use-case.

File: examples/chart/teleport/README.md
Comment (mumoshu) [reply examples/chart/teleport/README.md]: 👍

File: examples/chart/teleport/config/teleport.yaml
Comment (mumoshu) [reply examples/chart/teleport/config/teleport.yaml]: Good catch! I'll remove it

File: examples/chart/teleport/scripts/create-internal-pki
Comment (mumoshu) [reply examples/chart/teleport/scripts/create-internal-pki]: Okay then I'll move these scripts to my own repo, along with the steps to test Teleport locally on minikube 👍

Comment (mumoshu) [reply examples/chart/teleport/scripts/create-internal-pki]: > along with the steps to test Teleport locally on minikube

I'm saying this because `create-internal-pki` is a requisiet for `install-on-minikube`. The latter alone doesn't create a testable Teleport installation because we can't `tsh login` into it due to https://github.com/gravitational/teleport/pull/2014#issuecomment-399619312

Comment (mumoshu) [reply examples/chart/teleport/scripts/create-internal-pki]: Or do you want me to leave this script along with `install-on-minikube`?

File: examples/chart/teleport/minikube.values.yaml
Comment (mumoshu) [reply examples/chart/teleport/minikube.values.yaml]: Really?!

I had been believing that you're relying on golang's feature to respect the SSL_CERT_FILE envvar to laod the alternative CA cert as seen in https://github.com/golang/go/search?q=SSL_CERT_FILE&type=Code

AFAIK, there is no reference to TLS_CCERT_FILE in golang's code-base. Am I missing something?

File: examples/chart/teleport/config/teleport.yaml
Comment (mumoshu) [reply examples/chart/teleport/config/teleport.yaml]: Good point 👍  I was even considering to make the whole configuration provided as chart values, avoiding [the use of `{{ .Files.Get ... }}` in the template]({{ .Files.Get "config/teleport.yaml" | indent 4 }}). `Files.Get` works only for local charts. That's why you see [no `Files.Get`s for **user inputs** in the official Helm charts repository](https://github.com/kubernetes/charts/search?q=%22Files.Get%22&unscoped_q=%22Files.Get%22).

Comment (klizhentas) [reply examples/chart/teleport/config/teleport.yaml]: I agree, let's move configuration straight to configmap then, so we can tweak it with values, or you think whole configfile can be a parameter?

File: examples/chart/teleport/minikube.values.yaml
Comment (klizhentas) [reply examples/chart/teleport/minikube.values.yaml]: I mean SSL prefix refers to an obsolete layer, but if you want to keep for compat reasons I'm ok with that

Comment (mumoshu) [reply examples/chart/teleport/minikube.values.yaml]: I believe that I'm with you about the obsolete layer.

I'm just saying that rewriting `SSL_CERT_FILE` to `TLS_CERT_FILE` would make the example teleport installation unfunctional, because the support for the `TLS_CERT_FILE` is implemented nowhere?

File: examples/chart/teleport/config/teleport.yaml
Comment (mumoshu) [reply examples/chart/teleport/config/teleport.yaml]: Addressed in the last commit

**Comment by mumoshu** [comment]
@klizhentas This is almost done. I appreciate it if you could answer my questions above so that I can sort out the remainings 😉

**Comment by klizhentas** [comment]
k, ping me when you want me to do a second pass

**Comment by mumoshu** [comment]
@klizhentas This is fully ready for your review now 👍

File: examples/chart/teleport/README.md
Comment (klizhentas) [inline examples/chart/teleport/README.md]: let's make it reverse to match teleport's behavior:

TLS should be always on by default, to turn on TLS, users must set `--set proxy.tls.insecure-no-tls` flag

File: examples/chart/teleport/templates/deployment.yaml
Comment (klizhentas) [inline examples/chart/teleport/templates/deployment.yaml]: as I mentioned above, let's make it reverse to match teleport's semantics

**Review by klizhentas — COMMENTED** [review]
looks good, one last thing to fix before merge it to match semantics of `--insecure-no-tls` flag with the chart.

File: examples/chart/teleport/README.md
Comment (mumoshu) [reply examples/chart/teleport/README.md]: Thanks. I've addressed it in 9426272. If your whole purpose is just to enable TLS by default, it is a matter of setting `enabled: true` in the default `values.yaml`, hence the changes in the commit.

Regarding the configuration key, as myself being a normal helm user, I'd expect every feature toggle to be named `foo.enabled=[true|false]` rather than `proxy.tls.insecure-no-tls`. You may already know but let me show [how those toggles are named in the official helm charts](https://github.com/kubernetes/charts/search?utf8=%E2%9C%93&q=%22.enabled%22&type=).

However, you can still consider using `--set extraArgs[0]=--insecure-no-tls` you if prefer the config key to match the flag name!

Also note that an idiomatic helm chart should just work after installed without any additional values. On the other hand the teleport chart after my last commit does not. That's because you need to create tls assets in a secret and a configmap beforehand. You'll eventually want to make it work without the prerequisites. Perhaps generating tls assets automatically within the chart? Just my two cents.

**Comment by mumoshu** [comment]
@klizhentas Thanks for the review! TLS is turned on by default now. See my comment at https://github.com/gravitational/teleport/pull/2014/files#r199368476 for more context and details.

File: examples/chart/teleport/README.md
Comment (klizhentas) [reply examples/chart/teleport/README.md]: tls will get generated by default if you don't set the values in config file, so by default we should omit tls_ keys in the config file (so they will be auto generated) and turn on tls by default

**Comment by mumoshu** [comment]
Oh! Good to know. I'll add the default tls_ keys back, and reflect it into
README. Thanks again
for your feedback.
2018年7月2日(月) 11:58 Alexander Klizhentas <notifications@github.com>:

**Comment by mumoshu** [comment]
@klizhentas Addressed https://github.com/gravitational/teleport/pull/2014#discussion_r199375912. Would you mind taking a look once again? Thanks.

**Review by klizhentas — APPROVED** [review]
looks good to me, thanks, lets squash, rebase from master, and I will merge

**Comment by klizhentas** [comment]
I would prefer you squash it manually and write up and format a longer commit message mentioning the key design points rather than use github UI for this

**Comment by mumoshu** [comment]
Thanks. All right! I'll do it asap.

**Comment by mumoshu** [comment]
Done. Would you mind taking a look again?

**Comment by klizhentas** [comment]
ok to test

**Comment by klizhentas** [comment]
@mumoshu I've cherry-picked your commit and merged via https://github.com/gravitational/teleport/pull/2059 so closing this one. Thanks for all the great work!

**Pull request closed without merging**
